mods/cnc/maps/nod02b/nod02b.lua

```diff
@@ -0,0 +1,230 @@
+
+	OnAnyDamaged(Grd2ActorTriggerActivator, Grd2TriggerFunction)
+	Trigger.AfterDelay(Atk8TriggerFunctionTime, Atk8TriggerFunction)
+	Trigger.AfterDelay(Atk7TriggerFunctionTime, Atk7TriggerFunction)
+	Trigger.OnAllRemovedFromWorld(Atk4ActorTriggerActivator, Atk4TriggerFunction)
+	Trigger.OnAllRemovedFromWorld(Atk3ActorTriggerActivator, Atk3TriggerFunction)
+
+	Trigger.AfterDelay(0, getStartUnits)
+	Trigger.AfterDelay(0, setTriggerAtk5)
+	InsertNodUnits()
+end
+
+Tick = function()
+	if Nod.HasNoRequiredUnits()  then
+		if DateTime.GameTime > 2 then
+			GDI.MarkCompletedObjective(GDIObjective)
+		end
+	end
+
+	if GDI.HasNoRequiredUnits() then
+		Nod.MarkCompletedObjective(NodObjective2)
+	end
+
+	if DateTime.GameTime % DateTime.Seconds(1) == 0 and not Nod.IsObjectiveCompleted(NodObjective1) and CheckForBase(Nod, NodBaseBuildings) then
+		Nod.MarkCompletedObjective(NodObjective1)
+	end
+
+	if DateTime.GameTime % DateTime.Seconds(3) == 0 and Barracks.IsInWorld then
+		checkProduction(GDI)
+	end
+end
+
+CheckForBase = function(player, buildings)
+	local checked = { }
+	local baseBuildings = Map.ActorsInBox(Map.TopLeft, Map.BottomRight, function(actor)
+		if actor.Owner ~= Nod or Utils.Any(checked, function(bldng) return bldng.Type == actor.Type end) then
+			return false
+		end
+
+		local found = false
+		for i = 1, #buildings, 1 do
+			if actor.Type == buildings[i] then
+				found = true
+				checked[#checked + 1] = actor
+			end
+		end
+		return found
+	end)
+	return #baseBuildings >= 3
+end
+
+OnAnyDamaged = function(actors, func)
+	Utils.Do(actors, function(actor)
+		Trigger.OnDamaged(actor, func)
+	end)
+end
+
+getActors = function(owner, units)
+	local maxUnits = 0
+	local actors = { }
+	for type, count in pairs(units) do
+		local globalActors = Map.ActorsInBox(Map.TopLeft, Map.BottomRight, function(actor)
+			return actor.Owner == owner and actor.Type == type and not actor.IsDead
+		end)
+		if #globalActors < count then
+			maxUnits = #globalActors
+		else
+			maxUnits = count
+		end
+		for i = 1, maxUnits, 1 do
+			actors[#actors + 1] = globalActors[i]
+		end
+	end
+	return actors
+end
+
+checkProduction = function(player)
+	local Units = Map.ActorsInBox(Map.TopLeft, Map.BottomRight, function(actor)
+		return actor.Owner == player and actor.Type == UnitToRebuild
+	end)
+
+	if #Units < GDIStartUnits then
+		local unitsToProduce = GDIStartUnits - #Units
+		if Barracks.IsInWorld and unitsToProduce > 0 then
+			local UnitsType = { }
+			for i = 1, unitsToProduce, 1 do
+				UnitsType[i] = UnitToRebuild
+			end
+			Barracks.Build(UnitsType)			
+		end
+	end
+end
+
+getStartUnits = function()
+	local Units = Map.ActorsInBox(Map.TopLeft, Map.BottomRight, function(actor)
+		return actor.Owner == GDI
+	end)
+	Utils.Do(Units, function(unit)
+		if unit.Type == UnitToRebuild then
+			GDIStartUnits = GDIStartUnits + 1
+		end
+	end)
+end
+
+setTriggerAtk5 = function()
+	local Harv = Map.ActorsInBox(Map.TopLeft, Map.BottomRight, function(actor)
```

Comment:
This is pretty gross.  It would be better to suppress the harvester spawning (add `FreeActor: false` to the `Refinery: proc`) and then place/spawn the harvester manually.